File: ChallengesDev/Java/002-median-of-two-sorted-arrays/Solucion.java
/**
 * Solution to Challenge 002: Median of Two Sorted Arrays.
 * Optimized implementation using Binary Search over partitions.
 */
public class Solucion {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Ensure nums1 is the smaller array to optimize binary search
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length;
        int n = nums2.length;
        int low = 0, high = m;

        while (low <= high) {
            int partitionX = (low + high) / 2;
            int partitionY = (m + n + 1) / 2 - partitionX;

            // Values at the partition borders
            int maxLeftX = (partitionX == 0) ? Integer.MIN_VALUE : nums1[partitionX - 1];
            int minRightX = (partitionX == m) ? Integer.MAX_VALUE : nums1[partitionX];

            int maxLeftY = (partitionY == 0) ? Integer.MIN_VALUE : nums2[partitionY - 1];
            int minRightY = (partitionY == n) ? Integer.MAX_VALUE : nums2[partitionY];

            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                // Correct partition found
                if ((m + n) % 2 == 0) {
                    return ((double) Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2;
                } else {
                    return (double) Math.max(maxLeftX, maxLeftY);
                }
            } else if (maxLeftX > minRightY) {
                // We are too far right in nums1, move left
                high = partitionX - 1;
            } else {
                // We are too far left in nums1, move right
                low = partitionX + 1;
            }
        }

        throw new IllegalArgumentException("Input arrays are not sorted.");
    }
}

/**
 * Complexity Analysis:
 * Time: O(log(min(m, n)))
 * Space: O(1) since we don't use extra memory proportional to input.
 */

File: ChallengesDev/C++/009-quicksort-dual-pivot/solucion.cpp
#include <vector>
#include <algorithm>

/**
 * Implementation of Dual-Pivot Quicksort.
 */
class DualPivotSorting {
public:
    void sort(std::vector<int>& nums, int low, int high) {
        if (low < high) {
            int lp, rp;
            partition(nums, low, high, lp, rp);
            sort(nums, low, lp - 1);
            sort(nums, lp + 1, rp - 1);
            sort(nums, rp + 1, high);
        }
    }

private:
    void partition(std::vector<int>& nums, int low, int high, int& lp, int& rp) {
        if (nums[low] > nums[high]) {
            std::swap(nums[low], nums[high]);
        }

        int p1 = nums[low];
        int p2 = nums[high];

        int i = low + 1;
        int j = high - 1;
        int k = low + 1;

        while (k <= j) {
            if (nums[k] < p1) {
                std::swap(nums[k], nums[i]);
                i++;
            } else if (nums[k] >= p2) {
                while (nums[j] > p2 && k < j) {
                    j--;
                }
                std::swap(nums[k], nums[j]);
                j--;
                if (nums[k] < p1) {
                    std::swap(nums[k], nums[i]);
                    i++;
                }
            }
            k++;
        }
        i--;
        j++;

        std::swap(nums[low], nums[i]);
        std::swap(nums[high], nums[j]);

        lp = i;
        rp = j;
    }
};

/**
 * Complexity Analysis:
 * Time: O(n log n) average.
 * Space: O(log n) for the recursion stack.
 */
